slim_build_context.md

```diff
@@ -0,0 +1,162 @@
+      support. We could just implement the 90% cases and worry about corner cases
+      as they're reported, but then they're corner cases of the corner case of
+      this whole spec, which makes them hard to prioritize and leaves the feature
+      feeling buggy regardless.
+
+2. ##### Add a `docker_build` option `only_rebuild_on` - only files listed here will trigger builds, but the build still has access to the full build context.
+
+   Pros:
+
+   1. Simple implementation.
+
+   Cons:
+
+   1. The user has to explicitly list dependencies, which feels like boilerplate/
+      bookkeeping, repeating whatever's in the Dockerfile. Also, as is often the
+      case with such boilerplate, introduces a source of annoyance/user error,
+      in that this might not match the Dockerfile, either because the user fails
+      to keep it in sync with the Dockerfile, or simply typos.
+   2. A simple dependency list isn't very flexible, e.g., what if you want to
+      include go but exclude unit tests, or include js but exclude ts.
+   3. Introduces a distinction between "files we include in the build" and
+      "files we watch to trigger a build", making the user's required
+      mental model a bit more complicated.
+
+3. ##### Add `docker_build` options `ignore` and/or `only`, which filter the build context.
+
+   Pros:
+
+   1. Simple implementation.
+   2. Similar to #2, but more flexible - e.g., can include `internal` and
+      exclude `internal/**/*_test.go`.
+
+   Cons:
+
+   1. Risks feeling boilerplatey, as with option #2 above.
+   2. The flexibility over #2 is nice, but also kind of weird that we'd end up
+      having to do, e.g., `['**', '!package.json']` sometimes (and we don't have
+      good evidence people are very aware of this concept from Dockerfiles,
+      but maybe we can mostly mitigate that by advertising in our docs?). An
+      simpler, less flexible alternative would be `includes=[...], excludes=[...]`.
+      That might be flexible enough.
+
+4. ##### Allow `docker_build`'s context param to take either the string (directory)
+   it currently does, or a new `BuildContext` type that specifies a directory
+   and also watch/build rules.
+
+   Pros:
+
+   1. Simple implementation.
+   2. As a byproduct, might be a good replacement for `deps` in `custom_build`
+      (i.e., could provide better control over which files get fed to the command)
+      I'm not sure if this actually solves a problem, but it at least keeps
+      syntax more consistent between `docker_build` and `custom_build`
+
+   Cons:
+
+   1. Feels mostly like syntactic overhead on top of #3
+   2. Adds extensibility, but there seems to be little reason to add this extra
+      syntax when we're not sure it's the direction we want to go in or that
+      we'll need that extensibility, and it's easy to add later, as needed with
+      little wasted work.
+
+## Proposal:
```

Comment:
good spec! some things worth considering:

1) How will the user know if they wrote the pattern correctly? do we offer any facility to help them understand what the pattern matches? do we emit warnings if the pattern is suspicious? (e.g., the pattern doesn't match any files, or the pattern matches all files)

2) how do we expect users to debug problems with the image? with the old docker_build, you could simulate it with `docker build .`, but if they have ignores, there's no easy way to simulate that with the cli

The answer to both might be "we expect this to be a problem rarely/occasionally/all the time", but it would be good to say that explicitly

Reply:
> How will the user know if they wrote the pattern correctly? do we offer any facility to help them understand what the pattern matches? do we emit warnings if the pattern is suspicious? (e.g., the pattern doesn't match any files, or the pattern matches all files)

I could see false positive warnings being really annoying (e.g., if you are ignoring swap files that aren't always there, or you're sharing one set of patterns across multiple resources, and not all patterns apply to all resources). I'm not sure if I've seen any tool like this that gives good feedback on file patterns (e.g., gitignore, dockerignore), and I kind of expect that users are used to iterating on patterns with a very lame feedback cycle (works / doesn't work). There's potentially a better option here, but none is occurring to me at the moment.

> how do we expect users to debug problems with the image? with the old docker_build, you could simulate it with docker build ., but if they have ignores, there's no easy way to simulate that with the cli

At the risk of being pedantic, I want to reframe this question a bit: I'd expect most debugging of problems with the image would continue to occur via Tilt builds. I assume what you're asking about is that particular slice of debugging that builds images outside of Tilt to rule out Tilt bugs / configuration as a source of error. I don't think this feature changes much. We expect this feature to basically be used as an optimization (whether causing fewer builds to happen or including less info in the build context) and not have much impact on what the build actually does, except in the case of misconfiguration. If the build works outside of Tilt but not inside Tilt, that still points at a Tilt problem, as before, but now the realm of possible Tilt problems is slightly larger. Because it's simply an optimization, during debugging, this new configuration option can be eliminated as a source of error by commenting it out.

Of course, that all might just be rationalizing not having a solution to either problem because I can't think of any good ones, and I'm all ears if you have ideas.

Reply:
added to doc